test-data/typecheck/test_request.yml

```diff
@@ -14,4 +14,8 @@
             content: |
                 from django.db import models
                 class MyUser(models.Model):
-                    pass
\ No newline at end of file
+                    pass
+-   case: request_object_user_type
+    main: |
+        from django.http.request import HttpRequest
+        reveal_type(HttpRequest.user.tag)  # N: Revealed type is 'Union[Literal['abstract_base_user'], Literal['anonymous_user']]'
```

Comment:
Awesome! Now we need to check that `is_authenticated` can potentially check for this tag and discriminate types based on it.

Reply:
(I am still not sure it is possible, but we are experimenting here)

Reply:
Well, I think it can be somehow done in `mypy_django_plugin`, but I'll need more time to investigate it. I also noticed that such such piece of code:

```
from django.http.request import HttpRequest

def index(request: HttpRequest):
    reveal_type(request.user.tag)
    reveal_type(request.user)
    reveal_type(HttpRequest.user.tag)
(...)
```

returns this in mypy:

```
polls/views.py:6: note: Revealed type is 'Literal['abstract_base_user']'
polls/views.py:7: note: Revealed type is 'django.contrib.auth.models.User'
polls/views.py:8: note: Revealed type is 'Union[Literal['abstract_base_user'], Literal['anonymous_user']]'
```

but I haven't discovered why is so yet. Maybe type discrimination is already done somewhere earlier in plugin's runtime?

Reply:
We can probably ask this question in https://gitter.im/python/typing
Because I have not tried this or even seen this anywhere.

Reply:
I've asked a question a few days ago, but got no response yet :/

Reply:
@sobolevn should I still look for an answer or just rebase to squash commits and we're ready to merge?

Reply:
Hey @kszmigiel 
I wrote this based on what I understood from the mypy-documentation, might be something I'm missing, but I believe this is what we're looking for.

This type-checks and accessing `foo`/`bar` on the wrong branch in the if gives the expected errors.

```
from typing import Literal, Union

class Foo:
    is_two: Literal[False]
    bar = "heyy"

class Foo2:
    is_two: Literal[True]
    foo = "Ho"

Foos = Union[Foo, Foo2]

def f(f: Foos):
    if f.is_two:
        print(f.foo)
    else:
        print(f.bar)
```

So I'd think that we'd have to just annotate `is_anonymous` with the `Literal` types in AbstractBaseUser/AnonymousUser, and then it'd work.

Can you fix this or should I open up a new PR based on yours?

Reply:
Hey again @kszmigiel I opened #383 with the change I proposed here. I hope you're okay with me continuing on the work in this PR. :smil

Reply:
@bullfest sorry I had no time to respond to this. Again, great job!

Reply:
No worries, I had no expectation that you would respond in the hour between my comments, it was simply my curiosity and willingness to contribute that couldn't hold me back from opening a PR :sweat_smile: